Allow custom close delay for tooltips

bill-kerr opened the pull request:
Implements a custom close delay for tooltips. Please let me know if I should open an issue first.

~~There does seems to be some inconsistency between the current default [`TOOLTIP_COOLDOWN` (500ms)](https://github.com/adobe/react-spectrum/blob/d7cb26e27cac7b43f6fdff68ce4c10c1185d1bb3/packages/%40react-stately/tooltip/src/useTooltipTriggerState.ts#L18) and the default duration of the [`<OpenTransition>` exit (350ms)](https://github.com/adobe/react-spectrum/blob/d7cb26e27cac7b43f6fdff68ce4c10c1185d1bb3/packages/%40react-spectrum/overlays/src/OpenTransition.tsx#L41). I didn't want to go as far as trying to modify anything with that, but just wanted to point it out. Not entirely sure if they are meant to sync up, but tests need to wait for the transition to end before checking the DOM to make sure the element no longer exists.~~

~~If this is not a good fit for `<TooltipTrigger>`, I think we could still add it to `useTooltipTriggerState` so those of us that use these hooks as headless primitives could customize this behavior, but that would mean excluding this prop from [`SpectrumTooltipTriggerProps`](https://github.com/adobe/react-spectrum/blob/d7cb26e27cac7b43f6fdff68ce4c10c1185d1bb3/packages/%40react-types/tooltip/src/index.d.ts#L35).~~

The `closeDelay` prop is omitted from `SpectrumTooltipTriggerProps` as discussed below.

Usage:
```tsx
function Tooltip({state, ...props}) {
  let {tooltipProps} = useTooltip(props, state);

  return (
    <span {...mergeProps(props, tooltipProps)}>
      {props.children}
    </span>
  );
}

function TooltipTrigger(props) {
  let state = useTooltipTriggerState(props);
  let ref = React.useRef();

  let {triggerProps, tooltipProps} = useTooltipTrigger(props, state, ref);

  return (
    <span>
      <button aria-label="trigger" ref={ref} {...triggerProps}>
        {props.children}
      </button>
      {state.isOpen &&
        <Tooltip state={state} {...tooltipProps}>{props.tooltip}</Tooltip>}
    </span>
  );
}

function App() {
  return (
    <TooltipTrigger closeDelay={0} tooltip="Helpful information">
      Trigger
    </TooltipTrigger>
  );
}
```

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [X] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [X] Filled out test instructions.
- [X] Updated documentation (if it already exists for this component).
- [X] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Run the "custom close delay" describe block inside of `TooltipTrigger.test.js`.

## 🧢 Your Project:

<!--- Company/project for pull request -->

LFDanLu commented:
reopening to fix build

bill-kerr commented:
@LFDanLu I've fixed the lint errors and test...I think. It was passing for me locally previously.

LFDanLu commented:
Great, I'll close and reopen the PR again, not entirely sure why this CircleCI Pipeline issue is popping up

LFDanLu commented on packages/@react-types/tooltip/src/index.d.ts:
```diff
@@ -26,6 +26,12 @@ export interface TooltipTriggerProps extends OverlayTriggerProps {
    */
   delay?: number,
 
+  /**
+   * The delay time for the tooltip to close. [See guidelines](https://spectrum.adobe.com/page/tooltip/#Warmup-and-cooldown).
+   * @default 500
+   */
+  closeDelay?: number,
+
```
@bill-kerr out of curiosity, what is your use case for this configurable tooltip cooldown? I assume you just want to customize the cooldown for your own custom tooltip trigger and have that be the standard cooldown in your component library/app?

bill-kerr replied:
@LFDanLu Yup, exactly right. I'm building an app and using some React Aria primitives to build out reusable components and don't particularly like the behavior where the tooltip hangs around awhile after hovering off of the trigger. For my design system, the cooldown would be 0.

I understand if it doesn't make sense for React Spectrum, but it does seem in the spirit of React Aria/Stately.

bill-kerr commented:
@LFDanLu I've implemented the following changes:
1) Moved the tests for a custom close delay to a new `useTooltipTriggerState.test.js` file inside of Stately. I added a note that most of the tests for `useTooltipTriggerState` are in the `TooltipTrigger.test.js` file.
2) Updated the tests to no longer account for the transition cooldown from `<OpenTransition>`.
3) Omitted `closeDelay` from `SpectrumTooltipTriggerProps`.
4) Removed the example from the documentation for `<TooltipTrigger>`.

LFDanLu commented:
GET_BUILD

LFDanLu commented on packages/@react-stately/tooltip/test/useTooltipTriggerState.test.js:
```diff
@@ -0,0 +1,200 @@
+
+      // run the rest of the way through the timer and confirm that it has closed
+      // Note that closeDelay is less than the default close time of 500ms
+      act(() => jest.advanceTimersByTime(closeDelay / 2));
+      expect(onOpenChange).toHaveBeenCalledWith(true);
+    });
```
We haven't tested that advancing the rest of the time (aka the other half of `closeDelay`) would cause `onOpenChange` to fire with `false` and that the tooltip would disappear in this test.

bill-kerr replied:
👍

LFDanLu commented on packages/@react-stately/tooltip/test/useTooltipTriggerState.test.js:
```diff
@@ -0,0 +1,200 @@
+
+      // run the rest of the way through the timer and confirm that it has closed
+      // Note that closeDelay is more than the default close time of 500ms
+      act(() => jest.advanceTimersByTime(closeDelay / 2));
```
Perhaps this should be `TOOLTIP_COOLDOWN - closeDelay` to advance the rest of the time for the custom cooldown?

bill-kerr replied:
Yep, good catch

LFDanLu commented on packages/@react-stately/tooltip/test/useTooltipTriggerState.test.js:
```diff
@@ -0,0 +1,200 @@
+  let onOpenChange = jest.fn();
+
+  beforeAll(() => {
+    jest.useFakeTimers('legacy');
```
Think we don't need the legacy fake timers for these tests.

LFDanLu reviewed (changes requested):
Just some small test refinements that I noticed. Would you like to update the docs for `useTooltipTrigger` for `closeDelay` as well? If not, that is fine, the team can handle that in a followup

snowystinger commented on packages/@react-stately/tooltip/src/useTooltipTriggerState.ts:
```diff
@@ -101,7 +101,7 @@ export function useTooltipTriggerState(props: TooltipTriggerProps = {}): Tooltip
         globalCooldownTimeout = null;
         globalWarmedUp = false;
-      }, TOOLTIP_COOLDOWN);
+      }, closeDelay);
```
This should probably be `Math.max(TOOLTIP_COOLDOWN, closeDelay)`
I don't think there is a case where we'd like for the global cooldown to be 0.

The two values happened to be the same for us in Spectrum land, and so they've been conflated a little, not your fault. Cooldown is not the same thing as close delay.

Close delay is how long a tooltip stays open after moving off the target. Cooldown is how long, after being warmed up, until they require a warm up again. Warm up meaning that all subsequent tooltips appear immediately.

bill-kerr replied:
Done. Thanks for clarifying. I had a suspicion something like that was going on.

LFDanLu replied:
@snowystinger I wonder if we should introduce a new prop for setting the cooldown as well. That behavior feels pretty spectrum-y (forget if we've discussed this before)

snowystinger replied:
Yeah, had that thought as well. I think it's not as important to change though and mostly unnoticeable. So I'm content to leave it without a separate prop until someone asks for it with a compelling usecase

bill-kerr commented:
> Just some small test refinements that I noticed. Would you like to update the docs for `useTooltipTrigger` for `closeDelay` as well? If not, that is fine, the team can handle that in a followup

@LFDanLu Made adjustments and updated docs. I appreciate the feedback and willingness to incorporate this change!

LFDanLu commented:
GET_BUILD

LFDanLu reviewed (approved):
LGTM, thanks for the updates! Happy to merge this in and the team can discuss https://github.com/adobe/react-spectrum/pull/4128#discussion_r1126834915 separately if need be.
